community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/StringSchemaKey.java

```diff
@@ -0,0 +1,226 @@
+
+    private long entityId;
+    private boolean entityIdIsSpecialTieBreaker;
+
+    // TODO something better or?
+    // TODO this is UTF-8 bytes for now
+    byte[] bytes;
+
+    @Override
+    public void setEntityIdIsSpecialTieBreaker( boolean entityIdIsSpecialTieBreaker )
+    {
+        this.entityIdIsSpecialTieBreaker = entityIdIsSpecialTieBreaker;
+    }
+
+    @Override
+    public boolean getEntityIdIsSpecialTieBreaker()
+    {
+        return entityIdIsSpecialTieBreaker;
+    }
+
+    int size()
+    {
+        return ENTITY_ID_SIZE + bytes.length;
+    }
+
+    @Override
+    public long getEntityId()
+    {
+        return entityId;
+    }
+
+    @Override
+    public void setEntityId( long entityId )
+    {
+        this.entityId = entityId;
+    }
+
+    @Override
+    public void from( long entityId, Value... values )
+    {
+        this.entityId = entityId;
+        entityIdIsSpecialTieBreaker = false;
+        assertValidValue( values ).writeTo( this );
+    }
+
+    private TextValue assertValidValue( Value... values )
+    {
+        if ( values.length > 1 )
+        {
+            throw new IllegalArgumentException( "Tried to create composite key with non-composite schema key layout" );
+        }
+        if ( values.length < 1 )
+        {
+            throw new IllegalArgumentException( "Tried to create key without value" );
+        }
+        if ( !Values.isTextValue( values[0] ) )
+        {
+            throw new IllegalArgumentException(
+                    "Key layout does only support strings, tried to create key from " + values[0] );
+        }
+        return (TextValue) values[0];
+    }
+
+    @Override
+    public String propertiesAsString()
+    {
+        return asValue().toString();
+    }
+
+    @Override
+    public Value asValue()
+    {
+        return bytes == null ? Values.NO_VALUE : Values.utf8Value( bytes );
+    }
+
+    // TODO perhaps merge these lowest/highest methods into parent
+    @Override
+    public void initAsLowest()
+    {
+        bytes = null;
+        entityId = Long.MIN_VALUE;
+        entityIdIsSpecialTieBreaker = true;
+    }
+
+    @Override
+    public void initAsHighest()
+    {
+        bytes = null;
+        entityId = Long.MAX_VALUE;
+        entityIdIsSpecialTieBreaker = true;
+    }
+
+    private boolean isHighest()
+    {
+        return entityIdIsSpecialTieBreaker && entityId == Long.MAX_VALUE && bytes == null;
+    }
+
+    /**
+     * Compares the value of this key to that of another key.
+     * This method is expected to be called in scenarios where inconsistent reads may happen (and later retried).
+     *
+     * @param other the {@link StringSchemaKey} to compare to.
+     * @return comparison against the {@code other} {@link StringSchemaKey}.
+     */
+    int compareValueTo( StringSchemaKey other )
+    {
+        // TODO cover all cases of bytes == null and special tie breaker and document
+        if ( bytes != other.bytes )
```

Comment:
this is wrong: for example, comparing highest and lowest key now will say they are the same

Reply:
Omg, good catch!

Reply:
your future reader would be glad to know that there are specific values that considered outside of this method and also be curious why its so

Reply:
Just like the current readers then ;) Do you think javadoc would be enough?